clock_control_nrf_auxpll refactor to support nrf2 API

Hi-Im-David opened the pull request:
Refactor of the existing nrf auxpll driver to support the NRF2 API. Also added a test in nrf_clock_control to verify the auxpll.

Hi-Im-David committed:
dts: bindings: clock: add nordic,nrf-auxpll

Introduce Nordic NRF AUXPLL binding include.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
dts: vendor: nordic: nrf54h20 update canpll to use auxpll binding

Added new binding for macros of possible auxpll frequency settings.

Will be used in future products also.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
drivers: clock_control: refactor nrf_auxpll driver to nrf2

Refactor of previous clock_control_nrf_auxpll.c to use
the nrf2 clock control API

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
drivers: can: update clock control to use nrf2_auxpll

Auxpll driver has been refactored to use nrf2 API instead.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

Hi-Im-David committed:
tests: drivers: Update nrf2 clock control tests for nrf2_auxpll driver

Test added for new nrf2_auxpll driver. Frequency checks are
against known usecases of the auxpll (USB, CAN and AUDIO).
Also update test to be more modular, only running certain tests
if in the devicetree.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

bjarki-andreasen commented on drivers/clock_control/Kconfig.nrf2_auxpll:
Agreed, then you don't need to do the "or" part below :)
```
#if (defined(CONFIG_CLOCK_CONTROL_NRF2) || defined(CONFIG_CLOCK_CONTROL_NRF2_AUXPLL))
```

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_auxpll.c:
```diff
@@ -0,0 +1,254 @@
+#endif
+
+/* Check dt-bindings match MDK frequency division definitions*/
+BUILD_ASSERT(
+	NRF_AUXPLL_FREQ_DIV_MIN			== NRF_AUXPLL_FREQUENCY_DIV_MIN		&&
+	NRF_AUXPLL_FREQ_DIV_AUDIO_44K1	== NRF_AUXPLL_FREQUENCY_AUDIO_44K1	&&
+	NRF_AUXPLL_FREQ_DIV_USB24M		== NRF_AUXPLL_FREQUENCY_USB_24M		&&
+	NRF_AUXPLL_FREQ_DIV_AUDIO_48K	== NRF_AUXPLL_FREQUENCY_AUDIO_48K	&&
+	NRF_AUXPLL_FREQ_DIV_MAX			== NRF_AUXPLL_FREQUENCY_DIV_MAX,
+	"Different AUXPLL frequency definitions in MDK and devicetree bindings");
```
Also, I would make these individual BUILD_ASSERTs since they are "free" (they don't take up any memory in firmware)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_auxpll.c:
```diff
@@ -0,0 +1,254 @@
+	unsigned int wait = 0U;
+	nrf_auxpll_task_trigger(config->auxpll, NRF_AUXPLL_TASK_START);
+
+	do {
+		locked = nrf_auxpll_mode_locked_check(config->auxpll);
+		if (!locked) {
+			printk("AUXPLL not locked, waiting...\n");
+			k_msleep(AUXPLL_LOCK_WAIT_STEP_MS);
+			wait += AUXPLL_LOCK_WAIT_STEP_MS;
+		}
+	} while (wait < AUXPLL_LOCK_TIME_MAX_MS && !locked);
```
Check out the [`WAIT_FOR()`](https://github.com/zephyrproject-rtos/zephyr/blob/d8b37753061a9e0ace95204eccddee1f4b4c477b/include/zephyr/sys/util.h#L882) macro :)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_auxpll.c:
```diff
@@ -0,0 +1,254 @@
+};
+
+#define CLOCK_CONTROL_NRF_AUXPLL_DEFINE(n)													  \
+	BUILD_ASSERT(																			  \
```
Please split into separate BUILD_ASSERTs for readibility :)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_auxpll.c:
```diff
@@ -0,0 +1,254 @@
+				.range = DT_INST_ENUM_IDX(n, nordic_range),                                   \
+			},                                                                         	      \
+		.frequency = DT_INST_PROP(n, nordic_frequency),                                       \
+		.out_div = DT_INST_PROP(n, nordic_out_div),                                           \
```
Add a build assert for this one, it must be more than 0 from what I can see in the get_rate API implementation :)

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf2_auxpll.c:
```diff
@@ -0,0 +1,254 @@
+	.cancel_or_release = api_cancel_or_release_auxpll,
+};
+
+#define CLOCK_CONTROL_NRF_AUXPLL_DEFINE(n)													  \
+	BUILD_ASSERT(																			  \
+		DT_INST_PROP(n, nordic_frequency) == NRF_AUXPLL_FREQUENCY_DIV_MIN	 ||          	  \
+		DT_INST_PROP(n, nordic_frequency) == NRF_AUXPLL_FREQUENCY_AUDIO_44K1 ||          	  \
+		DT_INST_PROP(n, nordic_frequency) == NRF_AUXPLL_FREQUENCY_USB_24M	 ||          	  \
+		DT_INST_PROP(n, nordic_frequency) == NRF_AUXPLL_FREQUENCY_AUDIO_48K  ||          	  \
+		DT_INST_PROP(n, nordic_frequency) == NRF_AUXPLL_FREQUENCY_DIV_MAX,              	  \
+		"Invalid nordic,frequency value in DeviceTree for AUXPLL instance " #n);      		  \
+    static struct dev_data_auxpll data_auxpll##n    = { 									  \
+	.dev = DEVICE_DT_INST_GET(n), 														  	  \
+	}; 																					      \
+	static const struct clock_control_nrf_auxpll_config config##n = {                         \
+		.auxpll = (NRF_AUXPLL_Type *)DT_INST_REG_ADDR(n),                                     \
+		.ref_clk_hz = DT_PROP(DT_INST_CLOCKS_CTLR(n), clock_frequency),                       \
+		.ficr_ctune = DT_REG_ADDR(DT_INST_PHANDLE(n, nordic_ficrs)) +                         \
+			      DT_INST_PHA(n, nordic_ficrs, offset),                                       \
+		.cfg =                                                                                \
+			{                                                                                 \
+				.outdrive = DT_INST_PROP(n, nordic_out_drive),                                \
+				.current_tune = DT_INST_PROP(n, nordic_current_tune),	                      \
+				.sdm_off = DT_INST_PROP(n, nordic_sdm_disable),                               \
+				.dither_off = DT_INST_PROP(n, nordic_dither_disable),                         \
+				.range = DT_INST_ENUM_IDX(n, nordic_range),                                   \
+			},                                                                         	      \
+		.frequency = DT_INST_PROP(n, nordic_frequency),                                       \
+		.out_div = DT_INST_PROP(n, nordic_out_div),                                           \
+	};                                                                                        \
+                                                                                              \
+	DEVICE_DT_INST_DEFINE(n, clock_control_nrf_auxpll_init, NULL, &data_auxpll##n, &config##n,\
+			      PRE_KERNEL_1, CONFIG_CLOCK_CONTROL_INIT_PRIORITY,                    		  \
+			      &drv_api_auxpll);
```
Specifically, keep the `\` aligned with
```
												\
```
(copy it)

bjarki-andreasen commented on tests/drivers/clock_control/nrf_clock_control/src/main.c:
```diff
@@ -9,13 +9,17 @@
 #include <zephyr/drivers/clock_control/nrf_clock_control.h>
 #include <zephyr/kernel.h>
 #include <zephyr/ztest.h>
+#if defined(CONFIG_CLOCK_CONTROL_NRF2_AUXPLL)
+#include <zephyr/dt-bindings/clock/nrf-auxpll.h>
+#endif
```
include guards not neccesary :) It does not hurt to include them unconditionally

bjarki-andreasen commented on tests/drivers/clock_control/nrf_clock_control/src/main.c:
```diff
@@ -9,13 +9,17 @@
 	size_t clk_specs_size;
 };
 
+#if defined(CONFIG_CLOCK_CONTROL_NRF2)
```
This test only builds for nrf54h20 ATM, so this if guard should not be needed AFAIK

bjarki-andreasen reviewed (dismissed):
Really nice work! A few of comments but really nice addition :)

Hi-Im-David committed:
drivers: clock_control: update nrf2 lfosc_get_accuracy

Move lfosc_get_accuracy away from common library as
not all devices need this function.

Signed-off-by: David Jewsbury <david.jewsbury@nordicsemi.no>

bjarki-andreasen commented:
CI will fail, but atleast some of that is inhereted from the nrf2 refactor PR this one is based on :)

Hi-Im-David commented:
Waiting for https://github.com/zephyrproject-rtos/zephyr/pull/91187 to be merged

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -67,6 +79,61 @@ static int clock_control_nrf_auxpll_off(const struct device *dev, clock_control_
+	int ret = clock_control_nrf_auxpll_on(dev_data);
+
+	if (dev_data->notify) {
+
```
remove empty line

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -67,6 +79,61 @@ static int clock_control_nrf_auxpll_off(const struct device *dev, clock_control_
+}
+
+static int api_request_auxpll(const struct device *dev,
+			    const struct nrf_clock_spec *spec,
```
alignment is off, should be
```
static int api_request_auxpll(const struct device *dev,
			      const struct nrf_clock_spec *spec,
...
```

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -67,6 +79,61 @@ static int clock_control_nrf_auxpll_off(const struct device *dev, clock_control_
+}
+
+static int api_release_auxpll(const struct device *dev,
+			       const struct nrf_clock_spec *spec)
```
alignment is off

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -67,6 +79,61 @@ static int clock_control_nrf_auxpll_off(const struct device *dev, clock_control_
+}
+
+static int api_cancel_or_release_auxpll(const struct device *dev,
+				      const struct nrf_clock_spec *spec,
```
alignment is off

bjarki-andreasen commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -99,15 +166,21 @@ static enum clock_control_status clock_control_nrf_auxpll_get_status(const struc
 static int clock_control_nrf_auxpll_init(const struct device *dev)
 {
+	struct dev_data_auxpll *dev_data = dev->data;
+	static const struct onoff_transitions transitions = {
+		.start = onoff_start_auxpll,
+		.stop = onoff_stop_auxpll
+	};
```
please move static data outside of function, its placed in ROM anyway so it does not affect the stack, and the scope of the file is limited

bjarki-andreasen reviewed (dismissed):
Really nice work! A few nits (and a rebase) and it should be good :)

bjarki-andreasen reviewed (dismissed):
Let's gooo

bjarki-andreasen commented:
ping @masz-nordic :)

anangl commented on drivers/clock_control/clock_control_nrf_auxpll.c:
```diff
@@ -67,6 +79,60 @@ static int clock_control_nrf_auxpll_off(const struct device *dev, clock_control_
+	struct dev_data_auxpll *dev_data =
+		CONTAINER_OF(mgr, struct dev_data_auxpll, mgr);
+
+	dev_data->notify = notify;
+	int ret = clock_control_nrf_auxpll_on(dev_data);
+
+	if (dev_data->notify) {
+		dev_data->notify(&dev_data->mgr, ret);
+	}
```
There's no need to store `notify` in `dev_data`.
```suggestion
	int ret = clock_control_nrf_auxpll_on(dev_data);

	notify(&dev_data->mgr, ret);
```
